### PR TITLE
[opt](nereids) support pushing down aggr distinct through join

### PR DESCRIPTION
**Pull request by xzj7019**
### What problem does this PR solve?

support pushing down agg distinct through join on one side.

Issue Number: close #xxx

Related PR: #xxx

Problem Summary:

### Check List (For Author)

- Test 

    - [x] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No colde files have been changed.
        - [ ] Other reason 

- Behavior changed:

    - [x] No.
    - [ ] Yes. 

- Does this need documentation?

    - [x] No.
    - [ ] Yes. 

- Release note

    
    
    None

### Check List (For Reviewer who merge this PR)

- [x] Confirm the release note
- [x] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by xzj7019** [comment]
run buildall

**Comment by xzj7019** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+public class PushDownAggWithDistinctThroughJoinOneSide implements RewriteRuleFactory {
+    @Override
+    public List<Rule> buildRules() {
+        return ImmutableList.of(
+                logicalAggregate(logicalProject(innerLogicalJoin()))
+                        .when(agg -> agg.child().isAllSlots())
+                        .when(agg -> agg.child().child().getOtherJoinConjuncts().isEmpty())
+                        .when(agg -> !agg.isGenerated())
+                        .whenNot(agg -> agg.getAggregateFunctions().isEmpty())
+                        .whenNot(agg -> agg.child().children().stream().anyMatch(p -> p instanceof LogicalAggregate))
```
Comment (feiniaofeiafei) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java:60]: agg -> agg.child().children().stream().anyMatch(p -> p instanceof LogicalAggregate) , this is always false.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        Map<Slot, NamedExpression> leftSlotToOutput = new HashMap<>();
+        Map<Slot, NamedExpression> rightSlotToOutput = new HashMap<>();
+        if (isLeftSideAggDistinct) {
+            leftPushDownGroupBy.add((Slot) leftFuncs.get(0).child(0));
+            Builder<NamedExpression> leftAggOutputBuilder = ImmutableList.<NamedExpression>builder()
+                    .addAll(leftPushDownGroupBy);
+            leftFuncs.forEach(func -> {
+                Alias alias = func.alias("PDADT_" + func.getName());
+                leftSlotToOutput.put((Slot) func.child(0), alias);
+            });
```
Comment (feiniaofeiafei) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java:139]: map leftSlotToOutput value seems to be useless

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java (hunk 1 above)
Comment (xzj7019) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java:60]: thx, should be changed to .whenNot(agg -> agg.child()
                                .child(0).children().stream().anyMatch(p -> p instanceof LogicalAggregate))

**Review by feiniaofeiafei — APPROVED** [review]

**Comment by xzj7019** [comment]
run buildall

**Comment by xzj7019** [comment]
run cloud_p0

**Comment by xzj7019** [comment]
run cloud_p0

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+/**
+ * Push down agg function with distinct through join on only one side.
+ */
+public class PushDownAggWithDistinctThroughJoinOneSide implements RewriteRuleFactory {
+    @Override
+    public List<Rule> buildRules() {
+        return ImmutableList.of(
+                logicalAggregate(logicalProject(innerLogicalJoin()))
+                        .when(agg -> agg.child().isAllSlots())
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java:54]: agg.child().isAllSlots()
is this restriction too strong?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+                        .when(agg -> agg.child().child().getOtherJoinConjuncts().isEmpty())
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java:55]: "getOtherJoinConjuncts().isEmpty()"
how about remove this?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+                        .when(agg -> !agg.isGenerated())
+                        .whenNot(agg -> agg.getAggregateFunctions().isEmpty())
+                        .whenNot(agg -> agg.child()
+                                .child(0).children().stream().anyMatch(p -> p instanceof LogicalAggregate))
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java:59]: why only check the first child? how about agg.child().child(1)?

**Comment by englefly** [comment]
why it is called oneSide? it could push down through both sides

**Comment by xzj7019** [comment]
run cloud_p0

**Comment by xzj7019** [comment]
run buildall

**Comment by xzj7019** [comment]
run cloud_p0

**Comment by xzj7019** [comment]
run buildall

**Comment by xzj7019** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java (hunk 4 above)
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java:55]: nit: use `when` on `innerLogicalJoin()`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+                    leftAggOutputBuilder.build(), join.left());
+        } else {
+            rightPushDownGroupBy.add((Slot) rightFuncs.get(0).child(0));
+            Builder<NamedExpression> rightAggOutputBuilder = ImmutableList.<NamedExpression>builder()
+                    .addAll(rightPushDownGroupBy);
+            rightJoin = new LogicalAggregate<>(ImmutableList.copyOf(rightPushDownGroupBy),
+                    rightAggOutputBuilder.build(), join.right());
+        }
+
+        Preconditions.checkState(leftJoin != join.left() || rightJoin != join.right());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java:147]: add error message

**Review by morrySnow — DISMISSED** [review]

**Comment by xzj7019** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java (hunk 3 above)
Comment (xzj7019) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java:54]: aggr with distinct pushing down has special rewrite logic and restrict this case carefully by safety

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java (hunk 4 above)
Comment (xzj7019) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java:55]: leave this clean up task for the whole series of aggr pushing down rules

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java (hunk 6 above)
Comment (xzj7019) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggWithDistinctThroughJoinOneSide.java]: done

**Comment by xzj7019** [comment]
run cloud_p0

**Comment by xzj7019** [comment]
run buildall

**Comment by xzj7019** [comment]
run cloud_p0

**Comment by xzj7019** [comment]
run buildall

**Comment by xzj7019** [comment]
run p0

**Comment by xzj7019** [comment]
run p0

**Review by englefly — APPROVED** [review]

**Pull request merged**